watertap/tools/oli_api/flash.py

```diff
@@ -0,0 +1,494 @@
+# this list of conditions and the following disclaimer in the documentation and/or
+# other materials provided with the distribution.
+#
+# 3. Neither the name of OLI Systems, Inc. nor the names of any contributors to
+# the software made available herein may be used to endorse or promote products derived
+# from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+# OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
+# SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+# OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
+# TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+# EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+# You are under no obligation whatsoever to provide any bug fixes, patches, or upgrades to the
+# features, functionality or performance of the source code ("Enhancements") to anyone; however,
+# if you choose to make your Enhancements available either publicly, or directly to OLI Systems, Inc.,
+# without imposing a separate written license agreement for such Enhancements, then you hereby grant
+# the following license: a non-exclusive, royalty-free perpetual license to install, use, modify, prepare
+# derivative works, incorporate into other computer software, distribute, and sublicense such enhancements
+# or derivative works thereof, in binary and source code form.
+###############################################################################
+
+__author__ = "Oluwamayowa Amusat, Alexander Dudchenko, Paul Vecchiarelli"
+
+
+import logging
+
+import yaml
+
+from copy import deepcopy
+from itertools import product
+from pyomo.environ import units as pyunits
+
+from watertap.tools.oli_api.util.watertap_to_oli_helper_functions import (
+    get_oli_name,
+    get_charge,
+    get_charge_group,
+)
+from watertap.tools.oli_api.util.fixed_keys_dict import (
+    water_analysis_properties,
+    optional_properties,
+    input_unit_set,
+    output_unit_set,
+    stream_output_options,
+)
+
+_logger = logging.getLogger(__name__)
+handler = logging.StreamHandler()
+formatter = logging.Formatter(
+    "OLIAPI - %(asctime)s - %(levelname)s - %(message)s", "%H:%M:%S"
+)
+handler.setFormatter(formatter)
+_logger.addHandler(handler)
+_logger.setLevel(logging.DEBUG)
+
+# TODO: consider config for file_name for each writing method
+
+
+class Flash:
+    """
+    A class to execute OLI Cloud flash calculations, replacing and augmenting WaterAnalysis class functionality.
```

Comment:
Should we revise this to clarify that this class is meant to represent "Main Methods" from OLI Engine API? Probably don't need to mention "replacing and augmenting WaterAnalysis class functionality" here either. In one of my other comments, I asked why you packed water analysis inside of Flash, thinking that "Flash" represented the isothermal flash function from OLI, but now I realize you are using "Flash" to contain any of the "main methods" available via OLI (right?).

Reply:
Yes, the Flash class will include all the main methods from the API. Also - I have reworked each of these 3 classes (OLIApi, CredentialManager, and Flash) to remove a lot of the inconsistencies you noted. Those are relatively small changes compared to the major restructuring involved in this PR